dev/PushNotifications/PushNotificationChannel.cpp

```diff
@@ -48,18 +51,75 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
 
     winrt::event_token PushNotificationChannel::PushReceived(winrt::Windows::TypedEventHandler<winrt::Microsoft::Windows::PushNotifications::PushNotificationChannel, winrt::Microsoft::Windows::PushNotifications::PushNotificationReceivedEventArgs> handler)
     {
-        return m_channel.PushNotificationReceived([weak_self = get_weak(), handler](auto&&, auto&& args)
+        if (IsPackagedApp())
+        {
+            return m_channel.PushNotificationReceived([weak_self = get_weak(), handler](auto&&, auto&& args)
+            {
+                auto strong = weak_self.get();
+                if (strong)
+                {
+                    handler(*strong, winrt::make<winrt::Microsoft::Windows::PushNotifications::implementation::PushNotificationReceivedEventArgs>(args));
+                };
+            });
+        }
+        else
         {
-            auto strong = weak_self.get();
-            if (strong)
+            if (!m_isRegisteredWithLRP)
             {
-                handler(*strong, winrt::make<winrt::Microsoft::Windows::PushNotifications::implementation::PushNotificationReceivedEventArgs>(args));
-            };
-        });
+                wil::com_ptr<INotificationsLongRunningPlatform> notificationsLongRunningPlatform{
+                    wil::CoCreateInstance<NotificationsLongRunningPlatform, INotificationsLongRunningPlatform>(CLSCTX_LOCAL_SERVER) };
+
+                wchar_t processName[1024];
+                THROW_HR_IF(ERROR_FILE_NOT_FOUND, GetModuleFileNameExW(GetCurrentProcess(), NULL, processName, sizeof(processName) / sizeof(processName[0])) == 0);
+
+                THROW_IF_FAILED(notificationsLongRunningPlatform->RegisterForegroundActivator(this, processName));
+
+                m_isRegisteredWithLRP = true;
+            }
+            return m_foregroundHandlers.add(handler);
+        }
     }
 
     void PushNotificationChannel::PushReceived(winrt::event_token const& token) noexcept
     {
-        m_channel.PushNotificationReceived(token);
+        if (IsPackagedApp())
+        {
+            m_channel.PushNotificationReceived(token);
+        }
+        else
+        {
+            m_foregroundHandlers.remove(token);
+        }
```

Comment:
This implementation of unregistering the eventhandler has just leaked the long running foreground sink in the LRP. What you essentially need to do is to remove the foreground sink in LRP if the total number of event handlers registered is 0.

Reply:
Do we want the behavior to be unregister the LRP on 0 handlers then reregister on the next handler? 

Daniel and I were talking about the lifetime of the sink in the LRP and it is difficult to define when to remove it. One idea is to call UnregisterForegroundActivator when the PushNotificationChannel dtor is called, but the dtor will never be called because it still has a reference in the LRP. We could override AddRef/Release but I am not sure this is the way to go. 

Another idea is to have a cleanup function in the LRP that loops through its currently registered apps and checks their ref counts and removes them if their refcount == 1.

Any suggestions here?

Reply:
You remove it when the total number of foreground event handlers goes to 0. Otherwise it is ok to leak.

On the LRP side, when you try to invoke the FG Sink and you receive an RPC error, it is a good indication that the app is no longer in the FG. In that case, you will need to delete the foreground sink in the LRP and call into the background (protocol) activator. 

Makes sense? : )

Reply:
I would also remove it when the dtor of channel is called to be super safe.

Reply:
Yup sounds good!